### ISSUE TITLE
Generic_Sniffs_NamingConventions_ConstructorNameSniff: Wrong warning of usage of PHP4 style constructors when using namespaces

### ISSUE THREAD

**Issue by pmamat** [squizlabs/PHP_CodeSniffer#498]
test.php:
```
<?php
namespace Application\View\Helper
{
	class ApplicationBreadcrumbs extends \Zend_View_Helper_Abstract
	{
		public function ApplicationBreadcrumbs($breadcrumbs)
		{
			return 'Test!';
		}
	}
}

```

Output:
```
----------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------------------
 6 | ERROR | PHP4 style constructors are not allowed; use "__construct()" instead
----------------------------------------------------------------------------------
```

`ApplicationBreadcrumbs` is not a constructor of the class `Application\View\Helper\ApplicationBreadcrumbs`

**Comment by aik099** [comment]
That is expected behavior. The PHP4 style constructors are reported as error in all sniffs.

**Comment by aik099** [comment]
The methods are case-insensitive as PHP, so if that code would be executed on PHP4 then it would act as constructor.

**Comment by pmamat** [comment]
Is this really expected when using namespaces? 

As mentioned: this is no constructor.
When developing Zend View Helpers and using namespacing for class names the helper function needs to be named after the last part of the class name (class `Application_View_Helper_ApplicationBreadcrumbs` w/o namespaces, `Application\View\Helper\ApplicationBreadcrumbs` with namespaces): `applicationBreadcrumbs`.

I would expect a different behaviour (when using namespaces).

**Comment by pmamat** [comment]
Btw: when calling this function in PHP4 there is a parse error when using namespaces in PHP4:
```
Parse error:  syntax error, unexpected T_STRING in test.php on line 2
```

**Comment by pmamat** [comment]
(case-insensitive: yes, that has been a mistake, I corrected the example using the same casing in the name spaced class name and function)

**Comment by aik099** [comment]
Class constructor stays class constructor regardless of namespaces usage.

I believe in this case (method named as class was found in namespaced class) we should not raise the error as you've suggested based on fact, that PHP4 class constructor can't exist in class within namespace, that was introduced only in PHP5.3.

**Comment by pmamat** [comment]
Thank you for your reply!

I think we are in agreement:

1. **In the conventional class-only style (no namespace) no function may have the same name as the class**, since this is the PHP4 class constructor style (which is -- unfortunatly -- still accepted by PHP5).
2. **In a namespaced class this check is not necessary**, since the complete class name consists of the namespace + the given name (after `class`) separated by backslashes and consequently it is impossible to create a PHP4-style constructor class. Additionally there is the possibility to create a function named exactly as the "given name" (after `class`, without the namespace), which will not be recognized as a class constructor.

**Comment by gsherwood** [comment]
PHP_CodeSniffer does not track namespaces used throughout a file, so this sniff has no way of checking for the constructor name properly without affecting performance.

Given that this sniff's only purpose is to check for PHP5 code still using PHP4 constructors, and you are obviously using namespaces, you should just remove this sniff from your standard. I'll remove it from PHPCS itself sometime in the future (hopefully).

Is there a reason you need to keep this sniff?

**Comment by pmamat** [comment]
Greg, thanks for your response! I can see you point. When using namespaces, the sniff is not necessary any more.

We are having a project in which we are currently switching to the namespace style step by step, so we are having both namespaced and not-namespaced classes in our project. 

I would suggest to add a hint if this error occurs, that it can be ignored if namespaces are used.
